### DISCUSSION TITLE
Ainda precisamos do 'use strict'?

### DISCUSSION THREAD

**Discussion by felquis** [frontendbr/forum#300] [Discussão]
Continuando a discussão iniciada em https://github.com/frontendbr/forum/issues/257 porém com um pouco mais de acidez.
## Motivação

Se nosso código já está passando por um linter, que não nos permite fazer várias coisas erradas no JavaScript, por que precisamos do `'use strict'`?
## Motivação 2

Recentemente vi um projeto que tinha cerca de 260 ocasiões de `'use strict'`, isso me deixou pensando..

> Quantos bytes isso pesa?
## Motivação 3

Abri o código do Facebook, além de carregar 1.4mb de JS em 85 requests, ordenei a tabela pelo size dos arquivos, abri o maior que tinha e encontrei 680 ocorrências de `'use strict'` 
## Motivação 4

Teoricamente, não precisamos de centenas de 'use strict', mas se todo o código abaixo está sob nosso controle, precisamos de apenas 1 `'use strict'` visto que module bundlers, geralmente deixam todo o código embaixo de uma IIFE
## Como saber a quantidade de bytes que o use strict ocupa

O `use strict` pesa 12 bytes, pelo site [byte counter](https://mothereff.in/byte-counter#%27use%20strict%27) ou se vc salvar um arquivo com 10 `use strict` um do lado do outro, este arquivo irá pesar 121 bytes... 1 byte deve ser algo do arquivo mesmo.

```
"'use strict'".repeat(ocurrencies).length / 1024
```

Então deve ser isso mesmo, a quantidade de bytes é relacionada a quantidade de caracteres.. 

Então segundo a calculadora do Google bytes to kilobytes `680 * 12 / 1000` da 8.16kb
## Motivação 5

Se um site tem 1 milhão de unique page views no mês, e no Brasil temos sites com muito mais que isso.

(680 \* 12) \* 1000000 = 8160000000 o que da 8,16GB de transferência de dados apenas em 1 mês... e jogando isso por ano 8.16 \* 12 fácil fácil o Facebook se manter aquele script por 1 ano ele vai ter bem mais que 97GB de transferências só de `use strict`
## Conclusão

Precisamos mesmo do use strict se nosso projeto está bem lintado e sob o nosso controle? Se tivermos apenas 1 `use strict` já não deveria resolver?

NOTA: só neste artigo tem `"'use strict'".repeat(14).length * 12 / 1024` 1.96kb de use strict

**Comment by felquis** [comment]
## Motivação 6

O Google Analytics, provavelmente é o arquivo JavaScript mais usado no mundo! Sabem quantos use strict ele tem? https://www.google-analytics.com/analytics.js Nenhum. Provavelmente no Google, eles devem ter feito essa conta.

**Comment by felquis** [comment]
## Motivação 7

O Twitter, que é uma ótima referência diferente do Facebook, faz request de 4 arquivos JS + um native_bundle_v1 quando abri em aba privada.

jQuery + Twitter - init.ed8d97cfa822b126eec2277d2c5a8938053ff89a.js - 12 use strict
Do Twitter - timeline.4b5c939172f3a5f7406deac89d91d82695bfeafb.js - 4 use strict
Do Twitter - boot.e97a7d4740795306e2c10a6932c80c4d23eccedc.js - 2 use strict
De Terceiros - analytics.js - nenhum use strict
De Terceiros - native_bundle_v1_352605508afb747cf784ee7ba765b706ab7eff96.js - 62 use strict

**Comment by felquis** [comment]
O @grvcoelho que me questionou sobre isso a alguns dias, resolvi sair pesquisando e faz muito sentido.

Gostaria de saber a opinião de outras pessoas também.

**Comment by lagden** [comment]
no desenvolvimento tem que utilizar...

``` js
// 'use strict';

const a = [1, 2, 3];
for (y of a) {
    console.log(y);
}
```

**out:** `1, 2, 3`

``` js
'use strict';

const a = [1, 2, 3];
for (y of a) {
    console.log(y);
}
```

**out:** `ReferenceError: y is not defined`

e em produção que é o caso do GA e o seu também, deixa que o `minifyjs` tira tudo pra você!!!

Avoid bugs!! :feelsgood:

**Comment by grvcoelho** [comment]
@lagden: o primeiro ponto do @felquis foi a utilização de um linter!

Um linter ia pegar o erro do exemplo que você deu :D

**Comment by lagden** [comment]
sim... mas você não coda sozinho... se um outro cara pegar o código e for um novato, com certeza ele não terá o linter configurado...

**Comment by felquis** [comment]
@lagden pois é

> 'y' is not defined no-undef

http://eslint.org/docs/rules/no-undef

O Linter é um processo que passa no Continuous Integration Thiago.

**Comment by lagden** [comment]
o importante é deixar seu `dev code` showpeta com ou sem lint...
@felquis gosto de usar o XO: https://github.com/sindresorhus/xo

**Comment by fdaciuk** [comment]
Então, `'use strict'` é necessário para que o browser não faça as renderizações bizarras que a gente já conhece.. usando o `'use strict'` de forma global, você está dizendo que todo o código naquele escopo vai ser `strict`. Se tiver código não `strict`, vai dar cagada.

Pensando em ES6, teoricamente, quando você usa _ES6 Modules_, o seu código já é implicitamente `strict`, então com ES6 não precisa adicionar.

O Google Analytics provavelmente não usa porque não vai fazer diferença naquele código minúsculo - até porque não dá pra ter o Google como referência pra melhores práticas de código, ou estou errado? xD

E pra fechar, sobre a questão do linter, apesar de ele não permitir fazer coisas erradas, existem comportamentos que você não pode evitar: o `this` apontando para a `window` é um deles.

O ponto-e-vírgula ok, é desnecessário em boa parte dos casos, então podemos evitar, desde que o linter esteja configurado corretamente. Mas o `'use strict'` é mais questão de que "você controla o seu código, independente de como outros devs escreveram os seus".

**Comment by grvcoelho** [comment]
@lagden: não!

Linter é obrigatório! No nosso caso, tem um hook de precommit do github e passa no CI

Se o linter não for enforced, ele não serve para nada

**Comment by lagden** [comment]
mas o lint é configurável!!! muita atenção!!
eu lembro...
acho que um ano atrás...
o node 4 tinha um bug no `for of` que se você utilizar o `strict` ele funcionava correto e sem dava merda...

ele mesmo... node 4.0.0

``` js
// 'use strict';

const a = [1, 2, 3];
for (const y of a) {
    console.log(y);
}
```

_out:_ `1,1,1` 
e com `strict` `1, 2, 3`

lint foi pro espaço!!!

sempre digo... avoid bug!!!

**Comment by felquis** [comment]
Acredito que a moral da história é: Uma boa prática, é evitar dezenas de `use strict` no arquivo de produção.

**Comment by joaolucasl** [comment]
Visto que a grande maioria cria JS pro front com builds, module bundlers, etc, sou da opinião de que o `use strict` deve ser tratado do mesmo modo que flags de compilação de código nativa. No GCC, por exemplo tem as flags tipo `-pedantic` que tornam a checagem mais rigida quanto aos padrões. Passou da build step, dos linters, etc, o `use strict` é um pouco, senão totalmente, desnecessário.

**Comment by lagden** [comment]
@joaolucasl o mesmo bug se aplicava para o chrome já que ele utiliza o `V8` como engine... :feelsgood:

**Comment by joaolucasl** [comment]
Mas isso é algo bem pontual né @lagden? É um bug da engine, poderia mto ser o contrário: não funcionar no strict e funcionar no normal, ao que parece?

**Comment by lagden** [comment]
ai carai!! hihihi... :finnadie:

**Comment by joaolucasl** [comment]
To perguntando namoralzinha, fiquei curioso mesmo! Sem avacalhar hahahaha
Anyways, acho que ja contribui o que podia aqui, boa thread @felquis :)

**Comment by lagden** [comment]
@joaolucasl não é contigo...

o @felquis fez um lindo trabalho e a conclusão dele é excelente, mas pode dar problemas dependendo do **vendor**!!

exemplo: se você utiliza o código de um **vendor** e o pia escreveu assim...

``` js
var a = [1, 2, 3];
for (var i = 0; i < a.length; i++) {
    y = a[i];
}
```

sem o `use strict` vai funcionar mas esse código pode causar comportamentos estranho na sua aplicação e geralmente você não faz lint de vendor, presumo eu?!

colocar o **1** `use strict` no começo já iria parar tudo...

por isso recomendam utilizar o `strict` dentro do seu escopo...

o assunto é delicado...

**Comment by felquis** [comment]
Existe estes dois caras né: Código da Casa e Código de Terceiros.

No Código de Terceiro o máximo que fazemos é concatena-los em um bundle, da forma que eles são, ninguém altera um código de terceiro, eles permanecem intocáveis.

No Código da Casa é onde você tem total poder de edição, é aqui que aconcelho ter apenas 1 use strict.

Beleza @lagden?

**Comment by lagden** [comment]
@felquis 
se fica em arquivos separado então não precisa de **nenhum**

mas rola fazer coisas pretty crazy!!

**app.html**

``` html
<script src="bundle-vendor.js"></script>
<script src="bundle-my.js"></script>
```

**bundle-vendor.js**

``` js
var a = [1, 2, 3];
for (i = 0; i < a.length; i++) {
    y = a[i];
    console.log('vendor', y);
}
```

**bundle-my.js**

``` js
console.log(y);
var x = y || 1;
console.log(x);
```

**Comment by felquis** [comment]
No **bundle-my.js** o ESLint não vai deixar vc usar uma variável Y pois ela não foi definida.

Geralmente o **bundle-my.js** é algo como:

``` js
;(function (window, document, undefined) {
  'use strict'

  /* lots of js */
}(window, document))
```

**Comment by lagden** [comment]
sim... fiz um exemplo de qualidade altamente tosco!!

**Comment by rictorres** [comment]
eu tive uma problema em 2014 com relação a `'use strict'` :'(

meu time na época da comscore usava angular e um dos módulos que injetamos tinha `'use strict'` no global scope, o que quebrou o Highcharts.js que não suportava strict mode na época (não sei se já suporta).

sobre um dos seus argumentos... você esqueceu de `gzip`, não? 

```
.rw-r--r--  8.2k rictorres staff 10 Sep 18:34  use.strict

$ gzip use.strict

.rw-r--r--    76 rictorres staff 10 Sep 18:34  use.strict.gz
```

76 bytes ☺️

**Comment by felquis** [comment]
Muito bem notado @rictorres obrigado :D

**Comment by leobalter** [comment]
https://github.com/frontendbr/forum/issues/300#issuecomment-246025846

> O ponto-e-vírgula ok, é desnecessário em boa parte dos casos, então podemos evitar, desde que o linter esteja configurado corretamente. Mas o 'use strict' é mais questão de que "você controla o seu código, independente de como outros devs escreveram os seus".

Ponto e virgula é controlável se usar um minifier, irrelevante ter ou não, assim como quebra de linha.

ES Modules são use strict, sem declaração. Muitos desses códigos são gerados por transpiler que, para reproduzir um ambiente similar em ES5, colocam 'use strict' na função. Isso pode ser gerenciável via bundler (webpack, rollup, browserify, etc).

Usar o modo strict é muito mais questão de adaptação hoje em dia. Se você migrar seu código para ES Modules, você pode evitar uma dor de cabeça extra com erros surpresa. Mas aí é viver na incerteza: quando migrar.

Se você publica uma biblioteca para ser consumida por outras pessoas, é importante usar, como código de terceiros evita que encontrem incompatibilidades fora do seu escopo.

**Comment by suissa** [comment]
Eu não estou mais usando e nem ; por incrível que pareça LOL

Alguém aí está seguindo? https://github.com/feross/standard

**Comment by bernardodiasc** [comment]
Adiós `;`

;)

**Comment by lagden** [comment]
@suissa gosto do [XO](https://github.com/sindresorhus/xo)
eu não usava `;` em 2005! 😂 
mas parece uma boa idéia voltar aos velhos costumes!
